File: 64-minimum-path-sum/minimum-path-sum.cpp
class Solution {
public:
    
    int solve(vector<vector<int>> &grid, int m, int n, vector<vector<int>> &dp){
        if (m == 0 && n == 0) return grid[m][n];
        if (m < 0 || n < 0) return 100000;
        
        if (dp[m][n] != -1) return dp[m][n];

        int left = grid[m][n] + solve(grid, m, n-1, dp);
        int up = grid[m][n] + solve(grid, m-1, n, dp);

        return dp[m][n] = min(left, up);
    }
    int minPathSum(vector<vector<int>>& grid) {
        int m = grid.size();
        int n = grid[0].size();
        vector<vector<int>> dp(m, vector<int> (n, -1));
        return solve(grid, m-1, n-1, dp);
    }
};

File: 64-minimum-path-sum/minimum-path-sum.go
func minPathSum(grid [][]int) int {
    m := len(grid)
    n := len(grid[0])

    dp := make([]int, n)

    for i := 0; i < n; i++ {
        if i == 0 {
            dp[i] = grid[0][i]
            continue
        }
        dp[i] = grid[0][i] + dp[i-1]
    }

    for i := 1; i < m; i++ {
        tmp := make([]int, n)
        for j := 0; j < n; j++ {
            up := dp[j]
            left := math.MaxInt
            if j > 0 {
                left = tmp[j-1]
            }

            tmp[j] = grid[i][j] + min(up, left)
        }
        dp = tmp
    }

    return dp[n-1]
}